File: 123. Best Time to Buy and Sell Stock III.java
class Solution {
    public int maxProfit(int[] prices) {
        int len = prices.length;
        if (len < 1) {
            return 0;
        }
        
        int[] profit1 = new int[len];
        int[] profit2 = new int[len];
        profit1[0] = 0;
        profit2[len-1] = 0;
        
        int p1Min = prices[0], p2Max = prices[len - 1];
        
        for (int i = 1; i < len; ++i) {
            profit1[i] = Math.max(profit1[i - 1], prices[i] - p1Min);
            p1Min = Math.min(p1Min, prices[i]);
            
            int right = len - 1 - i;
            profit2[right] = Math.max(profit2[right + 1], p2Max - prices[right]);
            p2Max = Math.max(p2Max, prices[right]);
        }
        
        int maxProfit = 0;
        for (int i = 0; i < len; ++i) {
            maxProfit = Math.max(maxProfit, profit1[i] + profit2[i]);
        }
        
        return maxProfit;
    }
    
    // Solution 1: TLE
    public int maxProfit1(int[] prices) {
        
        int maxProfit = 0;
        for (int i = 0; i < prices.length; ++i) {
            int curProfit1 = maxProfixOneTransaction(prices, 0, i);
            int curProfit2 = maxProfixOneTransaction(prices, i+1, prices.length - 1);
            
            maxProfit = Math.max(maxProfit, curProfit1 + curProfit2);
        }
        
        return maxProfit;
    }
    
    private int maxProfixOneTransaction(int[] prices, int start, int end) {
        if (start > end || end > prices.length - 1) {
            return 0;
        }
        
        int bestBuy = prices[start], maxProfit = 0;
        
        for(int i = start + 1; i <= end; ++i) {
            if (prices[i] < bestBuy) {
                bestBuy = prices[i];
            }
            
            int curProfit = prices[i] > bestBuy ? prices[i] - bestBuy : 0;

            maxProfit = Math.max(maxProfit, curProfit);
        }
        
        return maxProfit;
    }
}

File: 123. Best Time to Buy and Sell Stock III.kt
class Solution {
    fun maxProfit(prices: IntArray): Int {
        val n = prices.size

        val firstMaxProfits: IntArray = IntArray(n)
        var minPrice = prices[0]
        var firstMaxProfit = 0

        for (i in 1 .. prices.size - 1) {
            if (prices[i] < minPrice ) {
                minPrice = prices[i]
            } else {
                firstMaxProfit = max(firstMaxProfit, prices[i] - minPrice)
            }

            firstMaxProfits[i] = firstMaxProfit
        }

        var secondProfits: IntArray = IntArray(n)
        var maxPrice = prices[n-1]
        var secondMaxProfit = 0
        for (i in prices.size - 2 downTo 0) {
            if (prices[i] > maxPrice) {
                maxPrice = prices[i]
            } else {
                secondMaxProfit = max(secondMaxProfit, maxPrice - prices[i])
            }

            secondProfits[i] = secondMaxProfit
        }

        var maxProfit = 0
        for (i in prices.indices) {
            maxProfit = max(maxProfit, firstMaxProfits[i] + secondProfits[i])
        }

        return maxProfit
    }
}
